xlsynth-g8r/src/equiv/prove_equiv.rs

```diff
@@ -1891,6 +2001,184 @@ mod test_utils {
+            #[should_panic]
+            #[test]
+            fn test_sel_unexpected_default_panics() {
+                let ir = r#"fn bad() -> bits[4] {
+                    idx: bits[2] = literal(value=1, id=1)
+                    a: bits[4] = literal(value=1, id=2)
+                    b: bits[4] = literal(value=2, id=3)
+                    c: bits[4] = literal(value=3, id=4)
+                    d: bits[4] = literal(value=4, id=5)
+                    def: bits[4] = literal(value=5, id=6)
+                    ret s: bits[4] = sel(idx, cases=[a,b,c,d], default=def, id=7)
+                }"#;
+                let f = crate::xls_ir::ir_parser::Parser::new(ir)
+                    .parse_fn()
+                    .unwrap();
+                let mut solver = <$solver_type>::new($solver_config).unwrap();
+                let inputs = get_fn_inputs(&mut solver, &f, None);
+                let _ = ir_to_smt(&mut solver, &inputs);
+            }
+        };
+    }
+
+    // one_hot_sel: multiple bits
+    #[macro_export]
+    macro_rules! test_one_hot_sel_multi {
+        ($solver_type:ty, $solver_config:expr) => {
+            crate::test_ir_fn_equiv!(
+                test_one_hot_sel_multi,
+                $solver_type,
+                $solver_config,
+                r#"fn f() -> bits[4] {
+                    sel: bits[3] = literal(value=3, id=1)
+                    c0: bits[4] = literal(value=1, id=2)
+                    c1: bits[4] = literal(value=2, id=3)
+                    c2: bits[4] = literal(value=4, id=4)
+                    ret o: bits[4] = one_hot_sel(sel, cases=[c0, c1, c2], id=5)
+                }"#,
+                r#"fn g() -> bits[4] {
+                    ret lit: bits[4] = literal(value=3, id=1)
+                }"#,
+                false,
+                EquivResult::Proved
+            );
+        };
+    }
+
+    // priority_sel: multiple bits -> lowest index wins
+    #[macro_export]
+    macro_rules! test_priority_sel_multi {
+        ($solver_type:ty, $solver_config:expr) => {
+            crate::test_ir_fn_equiv!(
+                test_priority_sel_multi,
+                $solver_type,
+                $solver_config,
+                r#"fn f() -> bits[4] {
+                    sel: bits[3] = literal(value=5, id=1)
+                    c0: bits[4] = literal(value=1, id=2)
+                    c1: bits[4] = literal(value=2, id=3)
+                    c2: bits[4] = literal(value=4, id=4)
+                    def: bits[4] = literal(value=8, id=5)
+                    ret o: bits[4] = priority_sel(sel, cases=[c0, c1, c2], default=def, id=6)
+                }"#,
+                r#"fn g() -> bits[4] {
+                    ret lit: bits[4] = literal(value=1, id=1)
+                }"#,
+                false,
+                EquivResult::Proved
+            );
+        };
+    }
+
+    // priority_sel: no bits set selects default
+    #[macro_export]
+    macro_rules! test_priority_sel_default {
```

Comment:
One general question -- could a lot of these be helper functions instead of macros? I think macros tend to negatively affect compile times when they're used pervasively so I think we may want to stick to functions when it's possible to do so.